echo-rest/src/main/java/com/netflix/spinnaker/echo/events/RestEventListener.java

```diff
@@ -55,79 +68,188 @@ class RestEventListener implements EventListener {
+
+      // only decorate sendEvent() to isolate the circuit breaker for POST calls only
+      circuitBreaker.executeRunnable(() -> sendEvent(eventMap, service));
+    } catch (CallNotPermittedException callNotPermittedException) {
+      if (log.isDebugEnabled()) {
+        // don't need to log this with INFO every time Circuit Breaker kicks in
+        handleError(event, service, callNotPermittedException);
+      }
+      log.error(
+          "Circuit Breaker '{}' blocked sending an event to {}",
+          circuitBreaker.getName(),
+          service.getConfig().getUrl(),
+          callNotPermittedException);
+    } catch (Exception e) {
+      handleError(event, service, e);
+    }
+  }
+
+  /**
+   * Handles errors that occur while processing or sending the event.
+   *
+   * @param event The event that caused the error.
+   * @param service The service to which the event was being sent.
+   * @param e The exception that occurred.
+   */
+  private void handleError(Event event, RestUrls.Service service, Exception e) {
+    if (event != null
+        && event.getDetails() != null
+        && event.getDetails().getSource() != null
+        && event.getDetails().getType() != null) {
+      log.error(
+          "Could not send event source={}, type={} to {}.\n Event details: {}",
+          event.getDetails().getSource(),
+          event.getDetails().getType(),
+          service.getConfig().getUrl(),
+          event,
+          e);
+    } else {
+      log.error("Could not send event.", e);
+    }
+
+    registry.counter("event.send.errors", "exception", e.getClass().getName()).increment();
+  }
+
+  /**
+   * Transforms the event into a map to be sent in the REST request.
+   *
+   * @param event The event to be transformed.
+   * @param service The service for which the transformation is done.
+   * @return The transformed event as a map.
+   */
+  private Map<String, Object> transformEventToMap(Event event, RestUrls.Service service) {
+    Map<String, Object> eventMap = null;
+    try {
+      eventMap = mapper.convertValue(event, new TypeReference<>() {});
+
+      if (service.getConfig().getFlatten()) {
+        eventMap.put("content", mapper.writeValueAsString(eventMap.get("content")));
+        eventMap.put("details", mapper.writeValueAsString(eventMap.get("details")));
+      }
+
+      if (service.getConfig().getWrap()) {
+        if (service.getConfig().getTemplate() != null) {
+          eventMap = restEventTemplateEngine.render(service.getConfig().getTemplate(), eventMap);
+        } else {
+          Map<String, Object> m = new HashMap<>();
+
+          m.put(
+              "eventName",
+              StringUtils.defaultString(service.getConfig().getEventName(), eventName));
+
+          m.put(StringUtils.defaultString(service.getConfig().getFieldName(), fieldName), eventMap);
+
+          eventMap = m;
+        }
+      }
+    } catch (Exception e) {
+      handleError(event, service, e);
+    }
+
+    return eventMap;
```

Comment:
if an exception is thrown in the `transformEventToMap` method. Is it possible that an empty event will be sent in the `sendEvent` method?
Not sure if that's the behavior we have previously.

Reply:
Good catch. Previously everything was in a single method and all the processing was done in a single try/catch block so if ObjectMapper threw an exception, the exception was logged and it never tried to send an empty event. I added a null check to avoid this scenario.